operator/rpc_client.go

```diff
@@ -2,421 +2,267 @@ package operator
+		return s1(submittedAt, err) && s2(submittedAt, err)
 	}
 }
 
-// Expected to be called with initialized client.
-func (c *AggregatorRpcClient) tryResendFromDeque() {
-	c.rpcClientLock.RLock()
-	defer c.rpcClientLock.RUnlock()
+type RetryFactory = func() RetryStrategy
 
-	c.unsentMessagesLock.Lock()
-	defer c.unsentMessagesLock.Unlock()
-
-	if len(c.unsentMessages) != 0 {
-		c.logger.Info("Resending messages from queue")
-	}
-
-	errorPos := 0
-	for i := 0; i < len(c.unsentMessages); i++ {
-		entry := c.unsentMessages[i]
-		message := entry.Message
-
-		// Assumes client exists
-		var err error
-		var reply bool
-
-		switch message := message.(type) {
-		case *messages.SignedCheckpointTaskResponse:
-			err = c.rpcClient.Call("Aggregator.ProcessSignedCheckpointTaskResponse", message, &reply)
-			if err != nil {
-				c.listener.IncErroredCheckpointSubmissions(true)
-			} else {
-				c.listener.IncCheckpointTaskResponseSubmissions(true)
-				c.listener.ObserveLastCheckpointIdResponded(message.TaskResponse.ReferenceTaskIndex)
-			}
-
-		case *messages.SignedStateRootUpdateMessage:
-			err = c.rpcClient.Call("Aggregator.ProcessSignedStateRootUpdateMessage", message, &reply)
-			if err != nil {
-				c.listener.IncErroredStateRootUpdateSubmissions(message.Message.RollupId, true)
-			} else {
-				c.listener.IncStateRootUpdateSubmissions(message.Message.RollupId, true)
-			}
-
-		case *messages.SignedOperatorSetUpdateMessage:
-			err = c.rpcClient.Call("Aggregator.ProcessSignedOperatorSetUpdateMessage", message, &reply)
-			if err != nil {
-				c.listener.IncErroredOperatorSetUpdateSubmissions(true)
-			} else {
-				c.listener.IncOperatorSetUpdateUpdateSubmissions(true)
-				c.listener.ObserveLastOperatorSetUpdateIdResponded(message.Message.Id)
-			}
-
-		default:
-			panic("unreachable")
-		}
-
-		if err != nil {
-			c.logger.Error("Couldn't resend message", "err", err)
-
-			if isShutdownOrNetworkError(err) {
-				c.logger.Error("Couldn't resend message due to shutdown or network error")
-
-				if errorPos == 0 {
-					c.unsentMessages = c.unsentMessages[i:]
-					return
-				}
-
-				for j := i; j < len(c.unsentMessages); j++ {
-					rpcMessage := c.unsentMessages[j]
-					c.unsentMessages[errorPos] = rpcMessage
-					errorPos++
-				}
-
-				break
-			}
-
-			entry.Retries++
-			if entry.Retries >= MaxRetries {
-				c.logger.Error("Max retries reached, dropping message", "message", fmt.Sprintf("%#v", message))
-				continue
-			}
-
-			c.unsentMessages[errorPos] = entry
-			errorPos++
-		}
-	}
-
-	c.unsentMessages = c.unsentMessages[:errorPos]
-	c.listener.ObserveResendQueueSize(len(c.unsentMessages))
+// By defaul, retry with a delay of 2 seconds between calls,
+// at most 10 times, and only if the error is recent enough (24 hours)
+// TODO: Discuss the "recent enough" part
+func DefaultAggregatorRpcRetry() RetryStrategy {
+	return RetryAnd(
+		RetryWithDelay(2*time.Second), RetryAnd(
+			RetryAtMost(10),
+			RetryIfRecentEnough(24*time.Hour)))
```

Comment:
What would be a reasonable TTL for messages? This is a very conservative default just to start the discussion.